Tests/Basic/OutputByteStreamTests.swift

```diff
@@ -25,30 +25,69 @@ class OutputByteStreamTests: XCTestCase {
         let streamable: Streamable = Character("!")
         stream.write(streamable)
 
-        stream.flush()
         
         XCTAssertEqual(stream.position, "Hello, world!".utf8.count)
+        stream.flush()
         XCTAssertEqual(stream.bytes, "Hello, world!")
     }
     
     func testStreamOperator() {
-        let stream = OutputByteStream()
+        let stream = BufferedOutputByteStream()
 
         let streamable: Streamable = Character("!")
         stream <<< "Hello" <<< Character(",") <<< Character(" ") <<< [UInt8]("wor".utf8) <<< [UInt8]("world".utf8)[3..<5] <<< streamable
         
         XCTAssertEqual(stream.position, "Hello, world!".utf8.count)
         XCTAssertEqual(stream.bytes, "Hello, world!")
 
-        let stream2 = OutputByteStream()
+        let stream2 = BufferedOutputByteStream()
         stream2 <<< (0..<5)
         XCTAssertEqual(stream2.bytes, [0, 1, 2, 3, 4])
     }
     
+    func testBufferCorrectness() {
+        let bytes1k = [UInt8](repeating: 0, count: 1 << 10)
+        let bytes4k = [UInt8](repeating: 0, count: 1 << 12)
+        let bytes8k = [UInt8](repeating: 0, count: 1 << 13)
+        let bytes16k = [UInt8](repeating: 0, count: 1 << 14)
+
+        var count = 0
+        let stream = BufferedOutputByteStream()
+
+        // Fill the buffer and add one more.
+        stream <<< bytes1k
+        stream <<< 0
+        count = bytes1k.count + 1
+        XCTAssertEqual(stream.bytes.count, count)
+
+        // Fill the buffer.
+        stream <<< bytes1k
+        count += bytes1k.count
+        XCTAssertEqual(stream.bytes.count, count)
+
+        // Fill the buffer in two halfs.
+        stream <<< bytes1k.prefix(upTo: bytes1k.count/2)
+        stream <<< bytes1k.prefix(upTo: bytes1k.count/2)
+        count += bytes1k.count
+        XCTAssertEqual(stream.bytes.count, count)
+
+        // Have something in buffer then add more past its size.
+        stream <<< bytes1k.prefix(upTo: bytes1k.count/2)
+        stream <<< bytes1k
+        count += bytes1k.count/2 + bytes1k.count
+        XCTAssertEqual(stream.bytes.count, count)
+
+        // Fill random.
+        stream <<< bytes1k <<< bytes4k <<< bytes8k <<< bytes16k <<< bytes16k <<< bytes8k <<< bytes4k <<< bytes1k
+        count += bytes1k.count + bytes4k.count + bytes8k.count + bytes16k.count + bytes16k.count + bytes8k.count + bytes4k.count + bytes1k.count
+
+        XCTAssertEqual(stream.bytes.count, count)
```

Comment:
I would probably have written this test a little differently, more like this:
```
// Check small / big interleavings for various block sizes (to prove buffer transfer optimizations).
for blockSize in [1 << 10, 1 << 12, 1 << 13, 1 << 14] {
  let smallBlock = ... make a 16 byte block ..
  let bigBlock == ... make a `blockSize` block ...
  XCTAssertEqual((BufferedOutputByteStream() <<< smallBlock <<< bigBlock).bytes, smallBlock + bigBlock)
  XCTAssertEqual((BufferedOutputByteStream() <<< bigBlock <<< smallBlock).bytes, smallBlock + bigBlock)
  XCTAssertEqual((BufferedOutputByteStream() <<< bigBlock <<< bigBlock).bytes, smallBlock + bigBlock)
}
```
which I think will cover the same stuff.